src/mtcp/mtcp_restart.c

```diff
@@ -858,42 +859,54 @@ static int read_one_memory_area(int fd)
       if (area.prot & PROT_WRITE) flags = O_RDWR;
     }
 
-    if (area.prot & MAP_SHARED) {
-      imagefd = mtcp_sys_open (area.name, flags, 0);  // Can we open file?
-      if (imagefd < 0 && mtcp_sys_errno == ENOENT) {
-        // File doesn't exist.  Do we have perm to create it and write data?
-        imagefd = mtcp_sys_open (area.name, O_CREAT|O_RDWR, 0);
-        if (imagefd >= 0) {
-          mtcp_sys_unlink(area.name);
-        } else {
-          // We don't have permission to re-create shared file.
-          // Open it as anonymous private, and hope for the best.
-          // FIXME: Or maybe we do have permission to re-create shared file,
-          // but it requires us to also re-create the parent directory.
-          area.flags ^= MAP_SHARED;
-          area.flags |= MAP_PRIVATE;
-          area.flags |= MAP_ANONYMOUS;
-          if (area.prot & PROT_WRITE) {
-            area.prot ^= PROT_WRITE;
-            MTCP_PRINTF("Found old underlying file %s\n"
-                        " that no longer exists, but it's mapped shared,"
-                        " with write permission,\n and we don't have permission"
-                        " to re-create it.  We will map it read-only\n"
-                        " and hope for the best.\n", area.name);
-          }
-          goto read_data;
-        }
-      }
+    int can_create_area = 1;
```

Comment:
I hate to do this to you, but could you explain why we moved the directory creation outside of `if (area.prot & MAP_SHARED) { }`? 

AFAIK, the directory tree matters only for the shared memory areas, so it's better to keep it all inside the 'if' block.